### PR TITLE
add DefaultArray type

### PR DESCRIPTION
**Pull request by simonbyrne**

### CODE REVIEW

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
I think this is ready. Any thoughts on name? I noticed `DeviceArray` was used elsewhere by @jkozdon and @lcw.

**Comment by lcw** [comment]
Does the new package loading system for CUDAnative and freinds https://discourse.julialang.org/t/cuarrays-cudanative-psa-simplified-package-loading/27897 change the way we want to select the default array type?

**Comment by simonbyrne** [comment]
@lcw that's a good point. It looks that it is a different mechanism, I will try to see if I can figure it out.

**Comment by simonbyrne** [comment]
I was unable to get that to get the CUDAapi query to work with precompilation: https://discourse.julialang.org/t/cuarrays-cudanative-psa-simplified-package-loading/27897/3?u=simonbyrne

@vchuravy do you have any suggestions?

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors ping

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
Okay, I've updated, also incorporating ideas from #316.

File: src/misc.jl [hunk 1]
```diff
@@ -1,11 +1,23 @@
...
-  
+function init_array_type()
+  MPI.Initialized() || MPI.Init()
+  if CuArrays.functional() && get(ENV, "CLIMA_GPU", "") != "false"
+    comm = MPI.COMM_WORLD
+    # allocate GPUs among MPI ranks
+    # this will oversubscribe GPUs, which is useful for testing
+    local_comm = MPI.Comm_split_type(comm, MPI.MPI_COMM_TYPE_SHARED,  MPI.Comm_rank(comm))
+    device!(MPI.Comm_rank(local_comm) % length(devices()))
+    CuArrays.allowscalar(x)
```
Comment (vchuravy) [inline src/misc.jl]: What is `x`?

File: src/misc.jl [hunk 2]
```diff
@@ -1,11 +1,23 @@
...
+GPUs can be explicitly disabled by setting the environment variable `CLIMA_GPU=false`.
 """
-haspkg(pkgname::String) = haskey(Pkg.installed(), pkgname)
-  
+function init_array_type()
+  MPI.Initialized() || MPI.Init()
+  if CuArrays.functional() && get(ENV, "CLIMA_GPU", "") != "false"
+    comm = MPI.COMM_WORLD
+    # allocate GPUs among MPI ranks
+    # this will oversubscribe GPUs, which is useful for testing
```
Comment (vchuravy) [inline src/misc.jl]: move this comment a line down?

File: src/CLIMA.jl [hunk 3]
```diff
@@ -1,6 +1,9 @@
 module CLIMA
 
-include("misc.jl")
+using CuArrays
+defaultarraytype() = CuArrays.functional() ? CuArray : Array
```
Comment (vchuravy) [inline src/CLIMA.jl]: Should this also check the environment flag like `init_default_array`?

File: src/Arrays/MPIStateArrays.jl [hunk 4]
```diff
@@ -529,30 +526,23 @@ using Requires
...
-  end
-  transform_cuarray(mpisa::MPIStateArray) = mpisa.realdata
-  transform_cuarray(x) = x
+# transform all arguments of `bc` from MPIStateArrays to CuArrays
+# and replace CPU function with GPU variants
+function transform_broadcasted(bc::Broadcasted, ::CuArray)
+  transform_cuarray(bc)
+end
+function transform_cuarray(bc::Broadcasted)
+  Broadcasted(CuArrays.cufunc(bc.f), transform_cuarray.(bc.args), bc.axes)
```
Comment (vchuravy) [inline src/Arrays/MPIStateArrays.jl]: I am not sure what this code intends to do, but this should probably be `GPUifyLoops.contextualize` and might also want to use `Adapt.jl`?

File: src/CLIMA.jl (hunk 3 above)
Comment (simonbyrne) [reply src/CLIMA.jl]: Ah, i meant to delete that.

File: src/misc.jl (hunk 1 above)
Comment (simonbyrne) [reply src/misc.jl]: ah, right.

File: src/Arrays/MPIStateArrays.jl (hunk 4 above)
Comment (simonbyrne) [reply src/Arrays/MPIStateArrays.jl]: Me either, I just un-indented it. Outside the scope of this PR, but feel free to address.

**Comment by simonbyrne** [comment]
This should work in theory, but is getting weird compile-time issues when CUDA is not present. Upstream issue: https://github.com/JuliaGPU/CUDAdrv.jl/issues/179

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
Blocked by
 - https://github.com/JuliaGPU/CuArrays.jl/issues/488
 - https://github.com/JuliaGPU/CuArrays.jl/issues/495

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

File: docs/make.jl [hunk 5]
```diff
@@ -47,8 +47,8 @@ makedocs(
       "AcceptableUnicode.md",
       "VariableList.md",
     ],
-    "Balance Law Examples" => ["BalanceLawOverview.md",
-                               GENERATED_BL_EXAMPLES...]
+    #"Balance Law Examples" => ["BalanceLawOverview.md",
+    #                           GENERATED_BL_EXAMPLES...]
```
Comment (charleskawczynski) [inline docs/make.jl]: Was this intentional?

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try-

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

File: docs/make.jl (hunk 5 above)
Comment (charleskawczynski) [reply docs/make.jl]: @simonbyrne

**Pull request merged**